### PR TITLE
[flang][hlfir] fix issue 118922

### PR DESCRIPTION
**Pull request by jeanPerier**
hlfir.elemental codegen optimize-out the final as_expr copy for temps local to its body, but sometimes, clean-up may have been emitted for this temp, and the code did not handle that.
This caused #118922 and @113843.

Only elide the copy if the as_expr is the last op.

### CODE REVIEW

File: flang/lib/Optimizer/HLFIR/Transforms/BufferizeHLFIR.cpp [hunk 1]
```diff
@@ -806,8 +806,13 @@ struct ElementalOpConversion
     // elemental is a "view" over a variable (e.g parentheses or transpose).
     if (auto asExpr = elementValue.getDefiningOp<hlfir::AsExprOp>()) {
       if (asExpr->hasOneUse() && !asExpr.isMove()) {
-        elementValue = hlfir::Entity{asExpr.getVar()};
-        rewriter.eraseOp(asExpr);
+        // Check that the asExpr is the final operation before the yield,
+        // otherwise, clean-ups could impact the memory being re-used.
+        mlir::Operation *nextOp = asExpr->getNextNode();
+        if (nextOp && nextOp == yield.getOperation()) {
```
Comment (tblah) [inline flang/lib/Optimizer/HLFIR/Transforms/BufferizeHLFIR.cpp:812]: nit: `yield.getOperation()` cannot return `nullptr` and so there is no need to check if `nextOp` is non-null.

It is okay if you prefer to keep the redundant check for clarity.

**Review by tblah — APPROVED** [review]
Thanks for the quick fix

**Review by vzakhari — APPROVED** [review]
LGTM

File: flang/lib/Optimizer/HLFIR/Transforms/BufferizeHLFIR.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Optimizer/HLFIR/Transforms/BufferizeHLFIR.cpp]: Good point, thanks!

**Pull request merged**